File: tests/grpc-benchmarks/swift-client/Sources/SwiftGrpcBenchmarkClient/UnaryBenchmarks.swift
/*
 * Copyright 2026 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
 */

import Foundation
import GRPCCore
import GRPCProtobuf

func unaryBenchmarks() -> [any Benchmark] {
    [
        unaryBenchmark(
            name: "unary-latency",
            description: "Warm sequential unary call latency",
            cases: [
                benchmarkCase(
                    warmupCalls: 100,
                    calls: 1_000,
                    concurrency: 1,
                    requestBytes: 0,
                    responseBytes: 0
                )
            ]
        ),
        unaryBenchmark(
            name: "unary-throughput",
            description: "Warm concurrent unary call throughput",
            cases: [
                benchmarkCase(
                    warmupCalls: 100,
                    calls: 10_000,
                    concurrency: 16,
                    requestBytes: 1_024,
                    responseBytes: 1_024
                )
            ]
        ),
        unaryBenchmark(
            name: "unary-payload-sweep",
            description: "Unary request and response payload size sweep",
            cases: payloadSweepCases()
        ),
        unaryBenchmark(
            name: "unary-concurrency-sweep",
            description: "Unary concurrency sweep with empty and 1 KiB payloads",
            cases: concurrencySweepCases()
        ),
    ]
}

private func unaryBenchmark(
    name: String,
    description: String,
    cases: [BenchmarkCase]
) -> any Benchmark {
    CallBenchmark(name: name, description: description, cases: cases) { client, parameters in
        let service = Grpc_Testing_BenchmarkService.Client(wrapping: client)
        let request = unaryRequest(
            requestBytes: parameters.requestBytes,
            responseBytes: parameters.responseBytes
        )

        return MeasuredCall(
            applicationBytes: Int64(parameters.requestBytes) + Int64(parameters.responseBytes),
            execute: {
                let response = try await service.unaryCall(request)
                guard response.payload.body.count == parameters.responseBytes else {
                    throw UnexpectedResponseSize(
                        expected: parameters.responseBytes,
                        actual: response.payload.body.count
                    )
                }
            }
        )
    }
}

private struct PayloadSweepSize {
    let name: String
    let bytes: Int
    let warmupCalls: Int
    let calls: Int
}

private let payloadSweepSizes = [
    PayloadSweepSize(name: "64b", bytes: 64, warmupCalls: 100, calls: 10_000),
    PayloadSweepSize(name: "1k", bytes: 1_024, warmupCalls: 100, calls: 10_000),
    PayloadSweepSize(name: "64k", bytes: 64 * 1_024, warmupCalls: 20, calls: 2_000),
    PayloadSweepSize(name: "1m", bytes: 1_024 * 1_024, warmupCalls: 5, calls: 200),
    PayloadSweepSize(
        name: "near-4m",
        bytes: 4 * 1_024 * 1_024 - 1_024,
        warmupCalls: 2,
        calls: 50
    ),
]

private func payloadSweepCases() -> [BenchmarkCase] {
    let symmetric = payloadSweepSizes.map { size in
        payloadCase(
            name: "symmetric-\(size.name)",
            size: size,
            requestBytes: size.bytes,
            responseBytes: size.bytes
        )
    }
    let upload = payloadSweepSizes.dropFirst().map { size in
        payloadCase(
            name: "upload-\(size.name)",
            size: size,
            requestBytes: size.bytes,
            responseBytes: payloadSweepAnchorBytes
        )
    }
    let download = payloadSweepSizes.dropFirst().map { size in
        payloadCase(
            name: "download-\(size.name)",
            size: size,
            requestBytes: payloadSweepAnchorBytes,
            responseBytes: size.bytes
        )
    }
    return symmetric + upload + download
}

private func payloadCase(
    name: String,
    size: PayloadSweepSize,
    requestBytes: Int,
    responseBytes: Int
) -> BenchmarkCase {
    benchmarkCase(
        name: name,
        warmupCalls: size.warmupCalls,
        calls: size.calls,
        concurrency: 1,
        requestBytes: requestBytes,
        responseBytes: responseBytes
    )
}

private func concurrencySweepCases() -> [BenchmarkCase] {
    [("empty", 0), ("1k", 1_024)].flatMap { payloadName, payloadBytes in
        [1, 2, 4, 8, 16, 32, 64, 128].map { concurrency in
            benchmarkCase(
                name: "\(payloadName)-c\(concurrency)",
                warmupCalls: max(100, concurrency * 10),
                calls: max(10_000, concurrency * 1_000),
                concurrency: concurrency,
                requestBytes: payloadBytes,
                responseBytes: payloadBytes
            )
        }
    }
}

private func benchmarkCase(
    name: String = "default",
    warmupCalls: Int,
    calls: Int,
    concurrency: Int,
    requestBytes: Int,
    responseBytes: Int
) -> BenchmarkCase {
    BenchmarkCase(
        name: name,
        parameters: try! BenchmarkParameters(
            warmupCalls: warmupCalls,
            calls: calls,
            concurrency: concurrency,
            requestBytes: requestBytes,
            responseBytes: responseBytes
        )
    )
}

private let payloadSweepAnchorBytes = 64

private func unaryRequest(requestBytes: Int, responseBytes: Int) -> Grpc_Testing_SimpleRequest {
    var payload = Grpc_Testing_Payload()
    payload.type = .compressable
    payload.body = Data((0..<requestBytes).lazy.map { UInt8($0 % 251) })

    var request = Grpc_Testing_SimpleRequest()
    request.responseType = .compressable
    request.responseSize = Int32(responseBytes)
    request.payload = payload
    return request
}

private struct UnexpectedResponseSize: Error, CustomStringConvertible, Sendable {
    let expected: Int
    let actual: Int

    var description: String {
        "Expected \(self.expected) response bytes, got \(self.actual)"
    }
}

File: tests/grpc-benchmarks/shared-kotlin-client/src/commonMain/kotlin/kotlinx/rpc/grpc/benchmarks/client/UnaryBenchmarks.kt
/*
 * Copyright 2026 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
 */

package kotlinx.rpc.grpc.benchmarks.client

import io.grpc.testing.integration.Payload
import io.grpc.testing.integration.PayloadType
import io.grpc.testing.integration.SimpleRequest
import io.grpc.testing.integration.invoke
import kotlinx.io.bytestring.ByteString

internal fun unaryBenchmarks(backend: BenchmarkBackend) = listOf(
    unaryBenchmark(
        backend = backend,
        name = "unary-latency",
        description = "Warm sequential unary call latency",
        cases = listOf(
            benchmarkCase(
                warmupCalls = 100,
                calls = 1_000,
                concurrency = 1,
                requestBytes = 0,
                responseBytes = 0,
            ),
        ),
    ),
    unaryBenchmark(
        backend = backend,
        name = "unary-throughput",
        description = "Warm concurrent unary call throughput",
        cases = listOf(
            benchmarkCase(
                warmupCalls = 100,
                calls = 10_000,
                concurrency = 16,
                requestBytes = 1_024,
                responseBytes = 1_024,
            ),
        ),
    ),
    unaryBenchmark(
        backend = backend,
        name = "unary-payload-sweep",
        description = "Unary request and response payload size sweep",
        cases = payloadSweepCases(),
    ),
    unaryBenchmark(
        backend = backend,
        name = "unary-concurrency-sweep",
        description = "Unary concurrency sweep with empty and 1 KiB payloads",
        cases = concurrencySweepCases(),
    ),
)

private fun unaryBenchmark(
    backend: BenchmarkBackend,
    name: String,
    description: String,
    cases: List<BenchmarkCase>,
): Benchmark {
    return CallBenchmark(name, description, cases, backend.implementationName) { client, parameters ->
        val unaryCall = backend.prepareUnaryCall(client)
        val request = benchmarkRequest(parameters.requestBytes, parameters.responseBytes)

        MeasuredCall(
            applicationBytes = parameters.requestBytes.toLong() + parameters.responseBytes,
            execute = {
                val response = unaryCall(request)
                check(response.payload.body.size == parameters.responseBytes) {
                    "Expected ${parameters.responseBytes} response bytes, got ${response.payload.body.size}"
                }
            },
        )
    }
}

private data class PayloadSweepSize(
    val name: String,
    val bytes: Int,
    val warmupCalls: Int,
    val calls: Int,
)

private val payloadSweepSizes = listOf(
    PayloadSweepSize("64b", 64, warmupCalls = 100, calls = 10_000),
    PayloadSweepSize("1k", 1_024, warmupCalls = 100, calls = 10_000),
    PayloadSweepSize("64k", 64 * 1_024, warmupCalls = 20, calls = 2_000),
    PayloadSweepSize("1m", 1_024 * 1_024, warmupCalls = 5, calls = 200),
    PayloadSweepSize(
        "near-4m",
        4 * 1_024 * 1_024 - 1_024,
        warmupCalls = 2,
        calls = 50,
    ),
)

private fun payloadSweepCases(): List<BenchmarkCase> = buildList {
    for (size in payloadSweepSizes) {
        add(payloadCase("symmetric-${size.name}", size, size.bytes, size.bytes))
    }
    for (size in payloadSweepSizes.drop(1)) {
        add(payloadCase("upload-${size.name}", size, size.bytes, PAYLOAD_SWEEP_ANCHOR_BYTES))
    }
    for (size in payloadSweepSizes.drop(1)) {
        add(payloadCase("download-${size.name}", size, PAYLOAD_SWEEP_ANCHOR_BYTES, size.bytes))
    }
}

private fun payloadCase(
    name: String,
    size: PayloadSweepSize,
    requestBytes: Int,
    responseBytes: Int,
): BenchmarkCase = benchmarkCase(
    name = name,
    warmupCalls = size.warmupCalls,
    calls = size.calls,
    concurrency = 1,
    requestBytes = requestBytes,
    responseBytes = responseBytes,
)

private fun concurrencySweepCases(): List<BenchmarkCase> = buildList {
    for ((payloadName, payloadBytes) in listOf("empty" to 0, "1k" to 1_024)) {
        for (concurrency in listOf(1, 2, 4, 8, 16, 32, 64, 128)) {
            add(
                benchmarkCase(
                    name = "$payloadName-c$concurrency",
                    warmupCalls = maxOf(100, concurrency * 10),
                    calls = maxOf(10_000, concurrency * 1_000),
                    concurrency = concurrency,
                    requestBytes = payloadBytes,
                    responseBytes = payloadBytes,
                ),
            )
        }
    }
}

internal fun benchmarkCase(
    name: String = "default",
    warmupCalls: Int,
    calls: Int,
    concurrency: Int,
    requestBytes: Int,
    responseBytes: Int,
): BenchmarkCase = BenchmarkCase(
    name = name,
    parameters = BenchmarkParameters(
        warmupCalls = warmupCalls,
        calls = calls,
        concurrency = concurrency,
        requestBytes = requestBytes,
        responseBytes = responseBytes,
    ),
)

private const val PAYLOAD_SWEEP_ANCHOR_BYTES = 64

internal fun benchmarkRequest(requestBytes: Int, responseBytes: Int): SimpleRequest {
    val body = ByteString(*ByteArray(requestBytes) { index -> (index % 251).toByte() })
    return SimpleRequest {
        responseType = PayloadType.COMPRESSABLE
        responseSize = responseBytes
        payload = Payload {
            type = PayloadType.COMPRESSABLE
            this.body = body
        }
    }
}
